### DISCUSSION TITLE
feat: add generation stats to `TextGeneration` trait

### DISCUSSION THREAD

**Discussion by erfanium** [TabbyML/tabby#860] [Ideas]
https://github.com/TabbyML/tabby/blob/99d49a9c43f6c0b68beed619cad0f8e24bb72d5d/crates/tabby-inference/src/lib.rs#L23-L31

Current TextGeneration trait is simple, but it doesn't tell the statics we need for monitoring and optimizing the inference server.

for example, `input_token_length` and `output_token_length` stats are really important to measure the inference server throughput.

My initial idea would be something like this:

```rs
pub struct TextGenerationStat {
    pub prompt_tokens_length: u32,
    pub output_tokens_length: Option<u32>, // not available in stream responses
    pub time: Option<u32>, // not available in stream responses
}

pub struct TextGenerationResponse {
    pub text: String,
    pub stat: Option<TextGenerationStat>,
}

pub struct TextGenerationStreamResponse {
    pub stream: BoxStream<String>,
    pub stat: Option<TextGenerationStat>,
}

pub trait TextGeneration: Sync + Send {
    async fn generate(
        &self,
        prompt: &str,
        options: TextGenerationOptions,
    ) -> TextGenerationResponse;
    async fn generate_stream(
        &self,
        prompt: &str,
        options: TextGenerationOptions,
    ) -> TextGenerationStreamResponse;
}
```

**Comment by wsxiaoys** [comment]
I think we can already compute stats for input string length / output string length / time. Should be sufficient for tabby's optimization purpose?

**Comment by erfanium** [comment]
@wsxiaoys yeah but as you know, `input string length != input token length`. metrics based on string length is not standard.

**Comment by wsxiaoys** [comment]
Yes - that's why I emphasis that `for tabby's optimization purpose` :).

**Comment by sundaraa-deshaw** [comment]
Is there a way (besides tabby server logs) to get a sense of inference performance (i.e. tokens/s) as part of the response for benchmarking? Since I am new to this space, can you please suggest other better alternatives to benchmark the server's performance?
